Introduce atom.directory-searcher service v0.1.0.

bolinfest opened the pull request:
The contract for a provider for the `atom.directory-searcher` service
is defined by the spec of the `DefaultDirectorySearcher`.

This modifies `Workspace::scan()` to use the appropriate `DirectorySearcher`
for each member of `atom.project.getDirectories()` when scanning the workspace
for files that match the specified regex.

bolinfest committed:
Introduce atom.directory-searcher service v0.1.0.

The contract for a provider for the `atom.directory-searcher` service
is defined by the spec of the `DefaultDirectorySearcher`.

This modifies `Workspace::scan()` to use the appropriate `DirectorySearcher`
for each member of `atom.project.getDirectories()` when scanning the workspace
for files that match the specified regex.

maxbrunsfeld commented on src/default-directory-searcher.coffee:
```diff
@@ -0,0 +1,61 @@
+    task.on 'scan:file-error', onSearchError
+    task.on 'scan:paths-searched', onPathsSearched
+
+    promise
```
I think if this interface is going to be a public API, we should probably find a way to avoid the `Promise that includes a cancel() method` thing. I know that's the way `Workspace::scan` behaves currently, but that's currently undocumented, and seems like a pattern we should prevent from spreading.

It might fit in better with Atom's other APIs if `::search()` returned a `DirectorySearch` object that allows the user to subscribe to different events:

```js
let search = searcher.search(theDirectory, "the-regexp", options);

search.onDidMatch((result) => console.log("Got result", result));
search.onDidError((error) => console.log("Error", error));
search.onDidSearchPaths((count) => console.log("Finished searching ", count, "files"));
search.onDidComplete((error) => console.log("Done"));

// ...

search.cancel();
```

@bolinfest @atom/core what do you think we should do for this API?

bolinfest replied:
@maxbrunsfeld I can't remove support for `cancel()` in this change because it is used in the `find-in-replace` package:

    @inProgressSearchPromise.cancel()

Also, I think it is valuable to continue to use a `Promise` to make it easier to compose with `Promise.all()`, use with the `await` keyword, etc.

At a minimum, I think it would have to continue to be a *thenable* such that is "Promise compatible," though I would have to test that out to be sure. (See http://www.html5rocks.com/en/tutorials/es6/promises/#toc-promise-terminology for info on *thenable*.)

bolinfest replied:
In terms of the callbacks, I think it would be cleaner to pass in a single object with an API similar to how delegates are often used in iOS (e.g., https://developer.apple.com/library/ios/documentation/UIKit/Reference/UIWebViewDelegate_Protocol/):

```
DirectorySearchDelegate {
  searcherDidMatch(result) {}
  searcherDidError(error) {}
  searcherDidSearchPaths(count) {}
  searcherDidComplete() {}
}
```

Then the caller passes in an instance of `DirectorySearchDelegate`, which avoids a bunch of event registration, deregistration. (I think.)

maxbrunsfeld replied:
> At a minimum, I think it would have to continue to be a thenable such that is "Promise compatible,"

That seems fine to me. I just think that since the returned object needs to have a `.cancel()` method, it would be clearer if were an instance of our own class, not of `Promise`.

> Then the caller passes in an instance of DirectorySearchDelegate, which avoids a bunch of event registration.

I see your point about the smaller amount of indirection involved when using the delegate pattern (one callback instead of potentially n callbacks), but in our situation I don't think there is much overhead or complexity involved in using event registration. All of Atom's existing APIs use the observer pattern, and it fits in well with our other tools like `CompositeDisposable`.

Also, I think it would be simpler implementation-wise as well, at least in the case of the `DefaultDirectorySearcher`. Rather than having to define a callback, like this:

maxbrunsfeld commented on src/workspace.coffee:
```diff
@@ -791,36 +800,72 @@ class Workspace extends Model
+      onPathsSearched = (provider, numberOfPathsSearched) ->
+        oldValue = numberOfPathsSearchedForProvider.get(provider)
+        if oldValue
+          totalNumberOfPathsSearched -= oldValue
```
What if a given `DirectorySearcher` object is being used for multiple searches simultaneously? Should this `Map` have `Directories` as keys, rather than `DirectorySearchers`?

bolinfest replied:
Good catch! Fixed.

maxbrunsfeld commented on src/workspace.coffee:
```diff
@@ -791,36 +800,72 @@ class Workspace extends Model
-    task = Task.once require.resolve('./scan-handler'), atom.project.getPaths(), regex.source, searchOptions, ->
-      deferred.resolve()
+    # Find a search provider for every Directory in the project.
+    providersAndDirectories = []
```
I think we should not use the term `provider` to refer to these objects. Probably just `searcher`.

bolinfest replied:
Changed to searcher throughout.

maxbrunsfeld commented on src/workspace.coffee:
```diff
@@ -791,36 +800,72 @@ class Workspace extends Model
+        onSearchError,
+        recordNumPathsSearched,
+        searchOptions))
+    searchPromise = Promise.all(allSearchPromises)
```
This means that when there are, for example 8 local root directories, we'll spawn 8 tasks to search them, rather than one task that searches all of them. This might be OK, but if we wanted to avoid it, we could de-dup the `atom.directory-searcher` services and make the `::search()` method take an *array* of `Directory` objects instead of just one `Directory`.

@atom/core @bolinfest do you think it matters how many tasks we spawn?

bolinfest replied:
My gut feeling is no. In general, I expect users will appreciate the reduction in latency.

Also, if this is a problem, we can fix this later by changing how the `DefaultDirectorySearcher` works. The `Promise` it returns could periodically `yield` to other active searchers, etc.

kevinsawicki commented on src/workspace.coffee:
```diff
@@ -829,11 +874,30 @@ class Workspace extends Model
+    # Although this method claims to return a `Promise`, the `ResultsPaneView.onSearch()`
+    # method in the find-and-replace package expects the object returned by this method to have a
+    # `done()` method. Include a done() method until find-and-replace can be updated.
+    cancellablePromise.done = (f) ->
```
what is `f`?  Mind giving it a more descriptive name?

bolinfest replied:
Renamed to `onSuccessOrFailure`.

kevinsawicki commented on src/workspace.coffee:
```diff
@@ -791,36 +800,72 @@ class Workspace extends Model
+    allSearchPromises = []
+    for entry in providersAndDirectories
+      {provider, directory} = entry
+      recordNumPathsSearched = onPathsSearched.bind(undefined, provider)
```
Seems a bit inconsistent to have `totalNumberOfPathsSearched` above but then shorten this to `recordNumPathsSearched` instead of `recordNumberOfPathsSearched`

bolinfest replied:
Changed to `recordNumberOfPathsSearched`.

kevinsawicki commented on src/workspace.coffee:
```diff
@@ -791,36 +800,72 @@ class Workspace extends Model
+      if providerForDirectory
+        providersAndDirectories.push({provider, directory})
+      else
+        throw Error("Could not find search provider for #{directory.getPath()}")
```
Since this methods return a promise, shouldn't the error be reported to that promise instead of thrown explicitly?

bolinfest replied:
I would say no: this is basically an assertion and we should fail fast. I wouldn't want this swept under the rug as an unhandled promise rejection.

If this error is thrown, we would have violated our invariant that there should always be a member of `@directorySearchers` for which `canSearchDirectory()` returns `true` for a given `Directory`, which we expect to uphold by having `DefaultDirectorySearcher` at the end of the list, as its `DefaultDirectorySearcher::canSearchDirectory()` indiscriminately returns `true`.

bolinfest committed:
Respond to comments from @maxbrunsfeld and @kevinsawicki.

bolinfest committed:
Change DirectorySearcher to return a DirectorySearch.

bolinfest commented:
@maxbrunsfeld I updated the code and tests so that a `DirectorySearcher` must return a `DirectorySearch` object that has the following methods:

* `then()`
* `onDidMatch()`
* `onDidError()`
* `onDidSearchPaths()`
* `cancel()`

Admittedly, I need to update all of the docs in `default-directory-search.coffee`, but I wanted to make sure you were OK with this API before I invested in that.

bolinfest committed:
Fix some nits I found during my self-review.

maxbrunsfeld commented:
Looks good to me. Thanks @bolinfest.

maxbrunsfeld commented on src/default-directory-searcher.coffee:
```diff
@@ -0,0 +1,91 @@
+    @task = new Task(require.resolve('./scan-handler'))
+    rootPaths = [directory.getPath()]
+    @promise = new Promise (resolve, reject) =>
+      myResolve = (arg) ->
```
for debugging?

bolinfest replied:
Good catch: fixed.

bolinfest committed:
Remove `myResolve` local variable that I was using for debugging.

maxbrunsfeld commented on spec/workspace-spec.coffee:
```diff
@@ -941,6 +942,100 @@ describe "Workspace", ->
+            resultPaths = []
+            cancelableSearch = atom.workspace.scan /aaaa/, ({filePath}) ->
+              resultPaths.push(filePath)
+            customDirectorySearcherPromiseInstance.cancel()
```
The *individual* `DirectorySearch` objects are not exposed to callers of `Workspace::scan`, right? Currently, this test is verifying the behavior of one of these individual searches. I think that instead, the test should assert is that calling `.cancel()` on the return value of `Workspace::scan` has the effect of calling `cancel` on each individual `DirectorySearch`.

bolinfest committed:
Two things:
1. Update documentation for `default-directory-search.coffee`.
2. Ensure that multiple `DirectorySearch` searches are run in series rather than in parallel to conserve resources.

bolinfest committed:
Remove empty param lists as per the linter.

bolinfest commented:
I just rebased because apparently my previous stack of diffs would not have merged cleanly because of the update to `spec/task-spec.coffee`.

bolinfest committed:
Switch to the delegate pattern,
eliminating a nasty race condition and deleting a lot of code.

bolinfest committed:
Make delegate method names more delegate-like.

bolinfest committed:
remove require for Disposable that is no longer needed

bolinfest committed:
Clean up test to verify that an individual failed search fails the overall search.

bolinfest committed:
Document and test the `cancel()` method on the object returned by `atom.workspace.scan()`.

bolinfest commented:
@maxbrunsfeld This fixes/amends the tests as we discussed.

bolinfest commented:
Rebased.

maxbrunsfeld commented on spec/workspace-spec.coffee:
```diff
@@ -941,6 +941,113 @@ describe "Workspace", ->
+                new CustomDirectorySearch(delegate)
+
+            atom.packages.serviceHub.provide(
+              "atom.directory-searcher", "0.1.0", new CustomDirectorySearcher())
```
I think these tests could be less verbose if you registered a searcher in a generic way in a `beforeEach`:

```coffee
fakeSearch = null

class FakeSearch
  constructor: (@delegate) ->
  then: (@simulateSuccess, @simulateFailure) ->
  cancel: -> @cancelled = true

beforeEach ->
  fakeSearch = null
  atom.packages.serviceHub.provide("atom.directory-searcher", "0.1.0", {
    canSearchDirectory: (directory) -> directory is dir1
    search: (directory, delegate) -> fakeSearch = new FakeSearch(delegate)
  })
```

Then the contents of the `it` blocks could be more focused on what's being tested:

```coffee
  it "can be cancelled...", ->
    # ...
    expect(fakeSearch.cancelled).toBe true

  it "fails if any particular search fails", ->
    # ...
    fakeSearch.simulateFailure("some-error");

  it "can override the DefaultDirectorySearcher...", ->
    # ...
    fakeSearch.delegate.didMatch({
      filePath: "the-path",
      matches: []
    })
```

bolinfest replied:
FYI, this is made difficult by:

```coffee
            waitsForPromise ->
              atom.workspace.scan /aaaa/, {onPathsSearched}, ({filePath}) ->
                resultPaths.push(filePath)
```

Because the `Promise` returned by `atom.workspace.scan` needs `fakeSearch` to do work in order for it to ever resolve.

maxbrunsfeld replied:
Yeah, I think you'd have to save off the return value of `scan`. So that you could manipulate the fake-search before waiting for the scan to complete.

bolinfest replied:
OK, I'm inserting this optional callback to make this work: you'll see.

maxbrunsfeld commented on spec/workspace-spec.coffee:
```diff
@@ -941,6 +941,113 @@ describe "Workspace", ->
+              cancelableSearch.then (promiseResult) -> resultOfPromiseSearch = promiseResult
+
+            runs ->
+              expect(resultOfPromiseSearch).toBe('cancelled')
```
Does this mean that if the search fails, then `.scan()` will always resolve with the string `"cancelled"`?

bolinfest replied:
Yes, though note this comment from workspace.spec:

```coffee
    # Make sure the Promise that is returned to the client is cancelable. To be consistent
    # with the existing behavior, instead of cancel() rejecting the promise, it should
    # resolve it with the special value 'cancelled'. At least the built-in find-and-replace
    # package relies on this behavior.
```

I would love to fix this in a subsequent diff, but it requires coordination between `find-and-replace` and this package.

maxbrunsfeld replied:
But that's only if the search is cancelled, right? The way I'm reading this test, it seems to assert that the promise is resolved with that value if a search fails for any reason.

bolinfest replied:
I assume it's OK to actually reject the thenable returned by `atom.workspace.scan()` in this case?

bolinfest replied:
On further consideration, this behavior might be hard to change because of the following:

```coffee
    cancellablePromise = new Promise (resolve, reject) ->
      onSuccess = ->
        resolve(null)
        return
      onFailure = ->
        resolve('cancelled')
        return
      searchPromise.then(onSuccess, onFailure)
```

I guess we want to say:
* If all the searches succeed, `cancellablePromise` resolves with `null`.
* If a search is cancelled, `cancellablePromise` resolves with `'cancelled'`.
* If a search fails, `cancellablePromise` rejects.

I think the tricky part is that it requires us to distinguish between a search that is cancelled vs. one that fails. This will probably require hoisting of resolve/reject functions...

bolinfest replied:
OK, I was able to do this without hoisting and the unit test now reflects this.

maxbrunsfeld commented on src/default-directory-searcher.coffee:
```diff
@@ -0,0 +1,107 @@
+  #   * `didError` {Function} call with an Error if there is a problem during the search.
+  #   * `didSearchPaths` {Function} periodically call with the number of paths searched thus far.
+  # * `options` {Object} with the following properties:
+  #   * `regexSource` {String} regex to search with. Produced via `RegExp::source`.
```
I think that `regexSource` should be a separate parameter, since it's so central to the point of the method. I would also say that the `delegate` and `options` objects could be merged. I don't think we should frame this like iOS/Cocoa, where the delegate is its own long-lived entity; it's just a collection of optional callbacks, which we're going to be creating for each call to `::search()`.

So I'd say a call should look like:

```coffee
search = searcher.search(theDirectory, "the-pattern", {
  ignoreCase: true,
  didError: ->
  didMatch: ->
  didSearchPaths: ->
})
```

bolinfest replied:
Point taken, will do.

bolinfest replied:
Though I think it should just get the raw `RegExp` and the searcher should decide whether it needs to invoke the `.source` param, no? I believe the only reason this was done originally is so it could be serialized over the process boundary, so that was leaking an implementation detail.

maxbrunsfeld replied:
> Though I think it should just get the raw RegExp and the searcher should decide whether it needs to invoke the .source param.

:+1: I like that.

maxbrunsfeld commented on src/default-directory-searcher.coffee:
```diff
@@ -0,0 +1,107 @@
+searchQueue = []
+nextId = 1
+activeSearchId = 0
+emitter = new EventEmitter
```
I'd like to avoid using global state here. I'm open to any alternatives you can think of, but my preference would be to just let `DirectorySearcher::search` take multiple directories, so we don't need queuing.

bolinfest replied:
OK, let me try multiple directory support. This should be easier to implement now that the return value of a `DirectorySearcher` is only required to have `then()` and `cancel()` as methods. (This was more difficult when it was also responsible for `onDidMatch`, `onDidPathsSearched`, etc.)

bolinfest replied:
Ugh, I think I still have to create yet-another-custom-class to achieve this. We'll see what you think.

bolinfest replied:
Actually, this may not be so bad.

bolinfest replied:
OK, you can see how I cleaned up `DefaultDirectorySearcher`, but I think I'm going to change the contract of `didSearchPaths` where the client must specify both the `Directory` and the count.

bolinfest replied:
All fixed.

maxbrunsfeld commented on src/workspace.coffee:
```diff
@@ -791,36 +800,76 @@ class Workspace extends Model
+        iterator(result) unless atom.project.isPathModified(result.filePath)
+      didError: (error) ->
+        iterator(null, error)
+    }
```
Was this to address some performance problem you were seeing in this method? It seems a little odd to use `Object.create` for this purpose. I would say just include these functions directly in the argument to `.search()` below, unless you're seeing that this speeds up the function significantly.

bolinfest replied:
OK, options is a much bigger object now, but I'll just create a new one each time.

maxbrunsfeld commented on src/workspace.coffee:
```diff
@@ -791,36 +800,76 @@ class Workspace extends Model
+      if searcher
+        searchersAndDirectories.push({searcher, directory})
+      else
+        throw Error("Could not find directory searcher for #{directory.getPath()}")
```
This is similar to what you did for the directory-provider right, where this exception is just here to signal an internal error? I was thinking it might be more clear and direct to explicitly choose the `DefaultDirectorySearcher` here, rather than putting it into the `@directorySearchers` array. Then we wouldn't need this `throw` here at all.

bolinfest replied:
Good call: will do.

bolinfest committed:
Responded to all of @maxbrunsfeld's comments except for:
* The effect of a failed search.
* Letting `DirectorySearcher::search` take multiple directories.

I'm working on those now.

bolinfest committed:
Changed DefaultDirectorySearcher to take multiple directories.
Still need to update `workspace.coffee` to make better use of this.

bolinfest committed:
Changed the contract of `DefaultDirectorySearcher` in two significant ways:
* `search()` takes an array of `Directory` objects rather than an individual object.
* `options.didSearchPaths` now takes the `Directory` in addition to the `count` as an argument.

bolinfest committed:
Changed the behavior so that if a searcher rejects,
then the thenable returned by `atom.workspace.scan()` rejects.

maxbrunsfeld commented on src/workspace.coffee:
```diff
@@ -46,6 +47,15 @@ class Workspace extends Model
+      'atom.directory-searcher',
+      '^0.1.0',
+      # New providers are added to the front of @directorySearchers because
+      # DefaultDirectorySearcher is a catch-all that will always claim to search a Directory.
```
I think you can :fire: this comment now.

bolinfest replied:
:fire:'d

maxbrunsfeld commented on src/workspace.coffee:
```diff
@@ -791,36 +801,67 @@ class Workspace extends Model
+      unless directories
+        directories = []
+        directoriesForSearcher.set(searcher, directories)
+      directories.push(directory)
```
Nice :+1:

bolinfest committed:
Comment excised, as recommended by @maxbrunsfeld.

maxbrunsfeld commented on src/default-directory-searcher.coffee:
```diff
@@ -0,0 +1,99 @@
+    isCancelled = false
+    promise = new Promise (resolve, reject) ->
+      run = ->
+        if isCancelled
```
I think you could simplify this even further: since the `scan-handler` task accepts multiple paths as arguments, you could just pass the whole array of directories into `DirectorySearch`, and have it pass all of their paths to the task.

bolinfest replied:
Does the task take care of making sure only one directory is searched at a time?

Also, I fear that might make it hard to implement `didSearchPaths` correctly since it is currently parameterized by `Directory`:

```coffee
    @task.on 'scan:paths-searched', (count) -> options.didSearchPaths(directory, count)
```

bolinfest replied:
Though if you want, I guess we can change the contract so that the first argument to `didSearchPaths` is the `provider` rather than the `directory`?

bolinfest replied:
OK, I just added two more commits that fix this.

Now, only one `scan-handler` task is created per call to `search()`.
I solved the `didSearchPaths` thing by passing each searcher its own `didSearchPaths` method.

bolinfest committed:
Two things:
* Removed the `Directory` argument to `didSearchPaths`.
Now each searcher gets its own instance of `didSearchPaths` that is parameterized by provider.
* Simplified `DefaultDirectorySearcher.search()` so it creates one `DirectorySearch`
rather than one per `Directory` passed to `search()`.

bolinfest committed:
Rename a variable to reflect a change in the previous commit.

bolinfest committed:
Update comment.

maxbrunsfeld commented:
Looks great. I'd say let's :ship: tomorrow.

bolinfest commented:
Thanks @maxbrunsfeld: sounds great!

kevinsawicki commented on src/workspace.coffee:
```diff
@@ -791,36 +799,65 @@ class Workspace extends Model
-  # Returns a `Promise`.
+  # Returns a *thenable* object with a `cancel()` method that will cancel all
+  # of the underlying searches that were started as part of this scan.
   scan: (regex, options={}, iterator) ->
```
This method still returns a `Promise` right? Can you change the doc back to a `Promise` from a `thenable` if it does?

bolinfest replied:
No, it now returns a *thenable*, which is an object with a Promise-compatible `then()` method.

A `Promise` has only two instance methods (http://www.html5rocks.com/en/tutorials/es6/promises/#toc-api):

* `then()`
* `cancel()`

I could also support `cancel`, I guess, but the returned object would not pass an `instanceof Promise` test because the code is:

```coffee
return {
  then: promise.then.bind(promise)
  cancel: ->
    isCancelled = true
    directorySearch.cancel()
}
```

In this case, it also supports an undocumented `done()` method, but as you can see in the code, that is only for backwards compatibility with `find-and-replace`.

kevinsawicki replied:
Looks like it returns this right: https://github.com/atom/atom/blob/3f03317b08647aaba5ef4c470af618a36ca1b76a/src/workspace.coffee#L874

Which is a `Promise`, am I missing something?

bolinfest replied:
Whoops, I went to the wrong file. Will fix.

bolinfest replied:
Fixed.

bolinfest committed:
Fix documentation bug.

bolinfest commented:
@kevinsawicki all good by you?

bolinfest commented:
@maxbrunsfeld @kevinsawicki I just rebased this on top of HEAD:master.

maxbrunsfeld commented:
:rocket: Thanks @bolinfest!

maxbrunsfeld commented:
Hey @bolinfest, we need to back this out for today's release, because it broke some specs in the `find-and-replace`. It looks like something to do with canceling? Can you take a look at these? We can re-merge shortly.

```
 ResultsModel
   cancelling a search
     it populates the model with all the results, and updates in response to changes in the buffer
       Expected spy unknown to have been called.
         at [object Object].<anonymous> (/private/var/lib/jenkins/workspace/atom/node_modules/find-and-replace/spec/results-model-spec.coffee:125:30)
         at process._tickCallback (node.js:367:9)
     it populates the model with all the results, and updates in response to changes in the buffer
       Expected spy unknown to have been called. (spec/results-model-spec.coffee:136:30)
```
